File: forms/code-of-conduct-notice/full-stack-with-loco-tera-htmx-alpine/src/engine/form_validator.rs
use super::flagged_issues::detect_additional_flags;
use super::types::{AssessmentData, FiredRule, GradingResult, ValidationStatus};
use super::utils::{completeness_percent, validation_status};
use super::validation_rules::{field_is_empty, validation_rules};

/// Pure function: grade a Code of Conduct Notice acknowledgement.
///
/// Returns a [`GradingResult`] containing:
/// - overall completeness percentage
/// - overall validation status (`Complete` / `Incomplete`)
/// - fired rules (one per unmet required field)
/// - additional flags (compliance-officer-facing alerts)
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = run_rules(data);
    let additional_flags = detect_additional_flags(data);

    let total_required = validation_rules().len() as u32;
    let completed_count = total_required - (fired_rules.len() as u32);
    let pct = completeness_percent(completed_count, total_required);
    let status = validation_status(pct);

    GradingResult {
        completeness_percent: pct,
        status,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

/// Evaluate every validation rule against the data, returning the rules
/// that fired (i.e. the required field is unmet).
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out = Vec::new();
    for rule in validation_rules() {
        if field_is_empty(data, rule.section, rule.field) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                section: rule.section.to_string(),
                description: rule.message.to_string(),
                field: rule.field.to_string(),
            });
        }
    }
    out
}

/// Convenience: return only the overall validation status (Complete / Incomplete).
pub fn overall_status(data: &AssessmentData) -> ValidationStatus {
    grade(data).status
}

File: forms/consent-to-treatment/front-end-with-html/js/form-validator.js
import { completenessPercent, validationStatus } from './types.js';
import { validationRules as rules } from './validation-rules.js';

// Pure form-completeness validator for the Consent To Treatment form.
// Mirrors `src/lib/engine/form-validator.ts`.
//
// Walks every rule from validation-rules.js, checks the corresponding
// field on the assessment data, and returns the completeness percent,
// status label, and the list of fired rules (missing required fields).

  

  /**
   * @param {import('./types.js').AssessmentData} data
   * @returns {{ completeness: number, status: 'Complete' | 'Incomplete', firedRules: import('./types.js').FiredRule[] }}
   */
  function validateForm(data) {
    const firedRules = [];
    

    for (const rule of rules) {
      const section = data[rule.section];
      const value = section ? section[rule.field] : undefined;
      if (value === '' || value === null || value === undefined) {
        firedRules.push({
          id: rule.id,
          section: rule.section,
          description: rule.message,
          field: rule.field
        });
      }
    }

    const totalRequired = rules.length;
    const completedCount = totalRequired - firedRules.length;
    const completeness = completenessPercent(completedCount, totalRequired);
    const status = validationStatus(completeness);

    return { completeness, status, firedRules };
  }

  

export { validateForm };
